src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1387 @@
+        finetune_i_bucket_main_stream = self._finetune_i_bucket_main_stream[:, :n_tokens, :n_tokens].repeat(batch_size,
+                                                                                                            1, 1)
+        finetune_i_bucket_predicting_stream = torch.cat([
+            self._finetune_i_bucket_predicting_stream[:, :n_tokens, :n_tokens],
+            self._finetune_i_bucket_predicting_stream[:, :n_tokens,
+            self.max_target_positions:self.max_target_positions + n_tokens]
+        ], 2).repeat(batch_size, 1, 1)
+        return finetune_i_bucket_main_stream, finetune_i_bucket_predicting_stream
+
+    def buffered_future_mask(self, tensor):
+        dim = tensor.size(0)
+        if not hasattr(self,
+                       '_future_mask') or self._future_mask is None or self._future_mask.device != tensor.device or self._future_mask.size(
+            0) < dim:
+            self._future_mask = torch.triu(fill_with_neg_inf(tensor.new(dim, dim)), 1)
+        return self._future_mask[:dim, :dim]
+
+    def buffered_future_mask_ngram(self, tensor):
+        dim = tensor.size(0)
+        if not hasattr(self,
+                       '_ngram_future_mask') or self._ngram_future_mask is None or self._ngram_future_mask.device != tensor.device:
+            self._ngram_future_mask = ngram_attention_bias(self.max_target_positions, self.ngram).type(tensor.dtype).to(
+                tensor.device)
+        ngram_future_mask = torch.cat([self._ngram_future_mask[:, :dim, :dim],
+                                       self._ngram_future_mask[:, :dim,
+                                       self.max_target_positions: self.max_target_positions + dim]
+                                       ], 2)
+        return ngram_future_mask
+
+    def forward(
+            self,
+            input_ids,
+            encoder_hidden_states,
+            encoder_padding_mask,
+            past_key_values=None,
+            use_cache=False,
+            output_attentions=False,
+            return_dict=False,
+            **unused,
+    ):
+        """
+
+        Args:
+            input_ids (LongTensor): previous decoder outputs of shape
+                `(batch, tgt_len)`, for teacher forcing
+            encoder_hidden_states: output from the encoder, used for
+                encoder-side attention
+            encoder_padding_mask: for ignoring pad tokens
+            decoder_cached_states (dict or None): dictionary used for storing state during generation
+            use_cache: inference or training procedure.
+
+        Returns:
+            tuple:
+                - the decoder's features of next n-grams, with shape `(batch, self.ngram * tgt_len, embed_dim)`
+                - hidden states
+                - attentions
+
+        """
+        if encoder_padding_mask is not None:
+            # remove bos to be consistent with fairseq version
+            encoder_padding_mask = encoder_padding_mask[:, 1:]
```

Comment:
Same here I think this should not be done. We could also just not add the bos token and would not have to cut it then